x-pack/filebeat/input/etw/input.go

```diff
@@ -0,0 +1,281 @@
+		once.Do(e.Close)
+	}()
+
+	wg.Wait() // Ensure all goroutines have finished before closing
+	close(errChan)
+	if err, ok := <-errChan; ok && err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// buildEvent builds the winlog object.
+func buildEvent(data map[string]any, h etw.EventHeader, session *etw.Session, cfg config) beat.Event {
+
+	winlog := map[string]any{
+		"activity_guid": h.ActivityId.String(),
+		"channel":       fmt.Sprintf("%d", h.EventDescriptor.Channel),
+		"event_data":    data,
+		"flags":         fmt.Sprintf("%d", h.Flags),
+		"keywords":      fmt.Sprintf("%d", h.EventDescriptor.Keyword),
+		"opcode":        fmt.Sprintf("%d", h.EventDescriptor.Opcode),
+		"process_id":    fmt.Sprintf("%d", h.ProcessId),
+		"provider_guid": h.ProviderId.String(),
+		"session":       session.Name,
+		"task":          fmt.Sprintf("%d", h.EventDescriptor.Task),
+		"thread_id":     fmt.Sprintf("%d", h.ThreadId),
+		"version":       fmt.Sprintf("%d", h.EventDescriptor.Version),
+	}
+
+	// Include provider GUID if not available in event header
+	zeroGUID := "{00000000-0000-0000-0000-000000000000}"
+	if winlog["provider_guid"] == zeroGUID {
+		winlog["provider_guid"] = session.GUID.String()
+	}
+
+	// Define fields map with Windows data and ECS mapping
+	fields := mapstr.M{
+		"winlog":         winlog,
+		"event.code":     fmt.Sprintf("%d", h.EventDescriptor.Id),
+		"event.created":  time.Now(),
+		"event.kind":     "event",
+		"event.severity": h.EventDescriptor.Level,
+	}
+
+	// Mapping from Level to Severity
+	levelToSeverity := map[uint8]string{
+		1: "critical",
+		2: "error",
+		3: "warning",
+		4: "information",
+		5: "verbose",
+	}
+
+	// Get the severity level, with a default value if not found
+	_, ok := levelToSeverity[h.EventDescriptor.Level]
+	if ok {
+		fields["log.level"] = levelToSeverity[h.EventDescriptor.Level]
+	}
+
+	// Include provider name if available
+	if cfg.ProviderName != "" {
+		fields["event.provider"] = cfg.ProviderName
+	}
+
+	// Include logfile path if available
+	if cfg.Logfile != "" {
+		fields["log.file.path"] = cfg.Logfile
+	}
+
+	return beat.Event{
+		Timestamp: convertFileTimeToGoTime(uint64(h.TimeStamp)),
+		Fields:    fields,
+	}
+}
+
+// convertFileTimeToGoTime converts a Windows FileTime to a Go time.Time structure.
+func convertFileTimeToGoTime(fileTime64 uint64) time.Time {
+	// Define the offset between Windows epoch (1601) and Unix epoch (1970)
+	const epochDifference = 116444736000000000
+	if fileTime64 < epochDifference {
+		// Time is before the Unix epoch, adjust accordingly
+		return time.Time{}
+	}
+
+	fileTime := windows.Filetime{
+		HighDateTime: uint32(fileTime64 >> 32),
+		LowDateTime:  uint32(fileTime64 & math.MaxUint32),
+	}
+
+	return time.Unix(0, fileTime.Nanoseconds())
+}
+
+// close stops the ETW session and logs the outcome.
+func (e *etwInput) Close() {
+	if err := e.operator.stopSession(e.etwSession); err != nil {
+		e.log.Error("failed to shutdown ETW session")
```

Comment:
I think you want a return after this statement otherwise, when the "logs the output" it will write both "failed to shutdown ETW session" and "successfully shutdown".